File: loggalitic/src/main/java/com/android/oleksandrpriadko/loggalitic/logger/Logger.java
package com.android.oleksandrpriadko.loggalitic.logger;

import com.android.oleksandrpriadko.loggalitic.policy.Policy;

/**
 * Created by Oleksandr Priadko.
 * 7/9/17
 */

public abstract class Logger {

    private Policy policy;

    public Logger(Policy policy) {
        this.policy = policy;
    }

    protected Policy getPolicy() {
        return this.policy;
    }

    public abstract boolean isLoggable(String tag, int level, boolean force);

    public abstract boolean isLogReportable(String tag, int level, boolean force);

    public abstract void v(String tag, String message, boolean force);

    public abstract void d(String tag, String message, boolean force);

    public abstract void i(String tag, String message, boolean force);

    public abstract void w(String tag, String message, boolean force);

    public abstract void e(String tag, String message, boolean force);

    public abstract void v(String tag, String message);

    public abstract void d(String tag, String message);

    public abstract void i(String tag, String message);

    public abstract void w(String tag, String message);

    public abstract void e(String tag, String message);

    public abstract void wtf(String tag, String message);

}

File: loggalitic/src/main/java/com/android/oleksandrpriadko/loggalitic/logger/Logger.kt
package com.android.oleksandrpriadko.loggalitic.logger

import com.android.oleksandrpriadko.loggalitic.policy.DefaultPolicy
import com.android.oleksandrpriadko.loggalitic.policy.Policy

abstract class Logger(protected val policy: Policy) {

    abstract fun isLoggable(tag: String, level: Int, force: Boolean): Boolean

    abstract fun isLogReportable(tag: String, level: Int, force: Boolean): Boolean

    abstract fun v(tag: String, message: String, force: Boolean)

    abstract fun d(tag: String, message: String, force: Boolean)

    abstract fun i(tag: String, message: String, force: Boolean)

    abstract fun w(tag: String, message: String, force: Boolean)

    abstract fun e(tag: String, message: String, force: Boolean)

    abstract fun v(tag: String, message: String)

    abstract fun d(tag: String, message: String)

    abstract fun i(tag: String, message: String)

    abstract fun w(tag: String, message: String)

    abstract fun e(tag: String, message: String)

    abstract fun wtf(tag: String, message: String)

    private class DummyLogger(policy: Policy) : Logger(policy) {

        override fun isLoggable(tag: String, level: Int, force: Boolean): Boolean {
            return false
        }

        override fun isLogReportable(tag: String, level: Int, force: Boolean): Boolean {
            return false
        }

        override fun v(tag: String, message: String, force: Boolean) {

        }

        override fun d(tag: String, message: String, force: Boolean) {

        }

        override fun i(tag: String, message: String, force: Boolean) {

        }

        override fun w(tag: String, message: String, force: Boolean) {

        }

        override fun e(tag: String, message: String, force: Boolean) {

        }

        override fun v(tag: String, message: String) {

        }

        override fun d(tag: String, message: String) {

        }

        override fun i(tag: String, message: String) {

        }

        override fun w(tag: String, message: String) {

        }

        override fun e(tag: String, message: String) {

        }

        override fun wtf(tag: String, message: String) {

        }
    }

    companion object {

        val NOT_SET: Logger = DummyLogger(DefaultPolicy())
    }

}
